# MitoNuclearCOEXPlorer
  * The MitoNuclearCOEXPlorer tool allows exploration of relationships between target gene(s) of nuclear origin and the mitochondrial genome  
  * Levaraging transcriptomic data from the GTEx project, nuclear-mitochondrial co-ordination was quantified by correlating expression levels of nDNA- and mtDNA-encoded genes  
  * The rationale here being that co-expressed genes may be directly regulating each other, controlled by the same regulatory regime, functionally related, members of the same pathway or part of the same protein complex, and as such, the correlation of their expression is meaningful  
  * See the [MitoNuclearCOEXPlorer](https://ainefairbrotherbrowne.shinyapps.io/MitoNuclearCOEXPlorer/) tool for analyses options offered and example outputs  

## Description
This repository contains the User Interface (UI) code for the MitoNuclearCOEXPlorer platform and accompanying back-end functions. 

## Getting Started
The tool can be reached here: [MitoNuclearCOEXPlorer](https://ainefairbrotherbrowne.shinyapps.io/MitoNuclearCOEXPlorer/)

## Related publication

## Developers
* **Aine Fairbrother-Browne** - UI and backend developer
  * Get in touch with Aine at [aine.fairbrother-browne.18@ucl.ac.uk](mailto:aine.fairbrother-browne.18@ucl.ac.uk).
* **Sonia Garc√≠a-Ruiz** - UI developer
  * Get in touch with Sonia at [s.ruiz@ucl.ac.uk](mailto:s.ruiz@ucl.ac.uk).

## Acknowledgments
All members of:  
* [Ryten Lab](https://snca.atica.um.es/RytenLab/)
* [Hodgkinson Lab](https://www.hodgkinsonlab.org/)
