# Mito-nuclear Brain Browser
The mito-nuclear brain browser tool allows exploration of relationships between target gene(s) of nuclear origin and the mitochondrial genome
Levaraging transcriptomic data from the GTEx project, nuclear-mitochondrial co-ordination was quantified by correlating expression levels of nDNA- and mtDNA-encoded genes. The rationale here being that co-expressed genes may be directly regulating each other, controlled by the same regulatory regime, functionally related, members of the same pathway or part of the same protein complex, and as such, the correlation of their expression is meaningful. See [https://snca.atica.um.es/mitonuclearbrainbrowser/] for analyses options offered by the mito-nuclear brain browser tool and example outputs.

## Description
This repository contains the User Interface (UI) code for the Mito-nuclear Brain Browser platform and accompanying backend functions. 

## Getting Started
**The Mito-nuclear Brain Browser can be reached following [https://snca.atica.um.es/mitonuclearbrainbrowser/] **

## Related publication

## Developers
* **Aine Fairbrother-Browne** - UI and backend developer
  * Get in touch with Aine at [aine.fairbrother-browne.18@ucl.ac.uk](mailto:aine.fairbrother-browne.18@ucl.ac.uk).
* **Sonia Garc√≠a-Ruiz** - UI developer
  * Get in touch with Sonia at [s.ruiz@ucl.ac.uk](mailto:s.ruiz@ucl.ac.uk).

## Acknowledgments
* All [RytenLab](https://snca.atica.um.es/RytenLab/Team) members
* All [HodgkinsonLab](https://www.hodgkinsonlab.org/) members
